### PR TITLE
Implement comments reporting

### PR DESCRIPTION
**Pull request by Feodor0090**
Adds "report" button.
Request scheme was written off chrome devtools by sending reports on dev server.
Form design was implemented just to work, i did not even trу to follow web design. Can try to make it closer on explicit request.
изображение

### CODE REVIEW

File: osu.Game/Overlays/Comments/ReportCommentPopover.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/ReportCommentPopover.cs:73]: I'm not sure where you got this behaviour from, but that's not how it works in web. "Commnets" section is required to submit a report.

File: osu.Game/Overlays/Comments/ReportCommentPopover.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/ReportCommentPopover.cs:46]: Use localised web strings instead (same applies everywhere a string is used for display in this PR, when possible). See `UsersStrings.Report*`.

File: osu.Game/Overlays/Comments/CommentReportReason.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/CommentReportReason.cs:13]: Add localised descriptions. See usages of `LocalisableDescriptionAttribute`.

File: osu.Game/Online/API/Requests/CommentReportRequest.cs
Comment (frenzibyte) [inline osu.Game/Online/API/Requests/CommentReportRequest.cs:37]: There is no existing documentation for the `reports` endpoint. Can you provide how you extracted the parameter names here, and how you found `comments` to be optional?

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/DrawableComment.cs:579]: Use lambda for single-line code like this.
```suggestion
        public Popover GetPopover() => new ReportCommentPopover(ReportComment);
```

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/DrawableComment.cs:424]: Not sure this should exist, but an OSD similar to the one in `ExternalLinkButton` may be better than a full notification.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Crashes when clicking "report" in beatmap info overlay.
```
[runtime] 2022-10-16 00:19:24 [verbose]: Unhandled exception has been denied .
Unhandled exception. System.InvalidOperationException: Cannot show or hide a popover without a parent PopoverContainer in the hierarchy
   at osu.Framework.Extensions.PopoverExtensions.setTargetOnNearestPopover(Drawable origin, IHasPopover target)
   at osu.Framework.Extensions.PopoverExtensions.ShowPopover(IHasPopover hasPopover)
   at osu.Game.Graphics.Containers.LinkFlowContainer.<>c__DisplayClass15_0.<createLink>b__0() in /Users/salman/Projects/osu/osu.Game/Graphics/Containers/LinkFlowContainer.cs:line 84
   at osu.Framework.Graphics.Containers.ClickableContainer.OnClick(ClickEvent e)
```

---

For the design, the popover looks too wide, and the dropdown has less space to contain long items (the popover also looks too spaced from the report button, which makes me not sure whether it's supposed to be a popover for the comment or a popover displayed on the centre of the screen). 

Generally, I don't think `Labelled*` components are a good choice for popovers, they're best suited for full screen areas like the editor setup screen where they're used.

I've tinkered around with this and I've came up with another `PopupDialog`-based design, but I can't say I'm a big fan of the ending result... (can pass code on request)

CleanShot 2022-10-16 at 03 38 46@2x

A custom implementation that matches web design might be best, rather than coming up with new ones.

File: osu.Game/Online/API/Requests/CommentReportRequest.cs
Comment (Feodor0090) [reply osu.Game/Online/API/Requests/CommentReportRequest.cs:37]: Parameter names:
изображение
Comment is requered to submit, my mistake.

**Comment by Feodor0090** [comment]
Made a form with web's design.
изображение
If it's okay, will and animations and push, but i'm not sure how to display it. `DialogOverlay` is made for this, but accepts only `PopupDialog`s. Should i inherit it and just overwrite all the content?

**Comment by frenzibyte** [comment]
I think a good approach would be to define an `IPopupDialog` interface publicly used by `DialogOverlay` and implement that in your custom implementation. Not entirely sure if that would work but would be an amicable one that avoids rewriting existing implementation or adding new local logic for it.

**Comment by Feodor0090** [comment]
WIP, waiting for #20796.

**Comment by Feodor0090** [comment]
Decided to use popover.

изображение

I updated it a little to look closer to web and resolved issues found during review. Could you check this again?

> can pass code on request

What did you use to create "Report Someone's comment?" string? "Report" string is from UsersStrings, i found it, but how to format username?

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/DrawableComment.cs:37]: The comment shouldn't be the popover target, that makes the popover display in completely unexpected positions:

https://user-images.githubusercontent.com/22781491/196821619-742ddc33-54f2-459f-86ad-7d443b5529da.mp4

The "Report" text should be the target. That should be possible by moving the text to a separate (`ReportCommentLabel`) component that inherits `IHasPopover` and is added to the flow via `AddArbitraryDrawable`.

File: osu.Game/Overlays/Comments/ReportCommentPopover.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/ReportCommentPopover.cs:59]: The dropdown shouldn't autosize the popover when opening it, that results in bad UX:

https://user-images.githubusercontent.com/22781491/196821976-532ce76a-d5eb-4890-bcb8-bcce4cd8879f.mp4

Should work similar to multiplayer match settings.

File: osu.Game/Overlays/Comments/ReportCommentPopover.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/ReportCommentPopover.cs:105]: Not sure if it's only me, but I would prefer button displaying in disabled state rather than shaking.

File: osu.Game/Overlays/Comments/ReportCommentPopover.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/ReportCommentPopover.cs:45]:
```suggestion
                        Text = ReportStrings.CommentTitle(comment.Comment.User?.Username ?? comment.Comment.LegacyName!),
```

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/DrawableComment.cs:593]: Looking at this again, maybe text placed in place of the "report" button would feel better than a toast:

https://user-images.githubusercontent.com/22781491/196823678-ceec56b6-5a01-4b3c-bb0c-a77a47835994.mp4

@peppy thoughts?

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Comments/DrawableComment.cs:421]: Not sure reporting a comment should hide all actions? One may want to take a permalink of the comment after reporting it, for example.

File: osu.Game/Overlays/BeatmapSetOverlay.cs
Comment (frenzibyte) [inline osu.Game/Overlays/BeatmapSetOverlay.cs:48]: Why is this not included in `CommentsContainer` instead?

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/BeatmapSetOverlay.cs
Comment (Feodor0090) [reply osu.Game/Overlays/BeatmapSetOverlay.cs:48]: When there is only one comment, there is not enough space for the popover:
изображение
I was not sure that it will work correctly in such cases, so made the whole overlay the container.

**Comment by frenzibyte** [comment]
I've pushed several commits to clean up the code and keep things self-contained, along with changes to the popover design. Feels quite polished now (save for the popover size compared to the comment, which can't be helped without resolving the [bigger issue at hand](https://github.com/ppy/osu/issues/17940) first).

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (frenzibyte) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: The text proposal has already been pushed in this PR, but I'll wait for further opinions before merge.

**Review by frenzibyte — DISMISSED** [review]

File: osu.Game/Overlays/Changelog/ChangelogContent.cs
Comment (peppy) [inline osu.Game/Overlays/Changelog/ChangelogContent.cs:14]: I'm really not sure about having to add these all over the place. Did we discuss having a global one which can be the fallback case for whenever a local one is not required? Is there an issue with doing that?

The reason I bring this up is that this weird interaction happens due to the local `PopOverContainer` being padding too close:

https://user-images.githubusercontent.com/191335/197678961-255a6418-e1ae-43ac-9316-8b035b89d739.mp4

**Review by peppy — COMMENTED** [review]
Code looks pretty great apart from the pointed out issue.

File: osu.Game/Overlays/Changelog/ChangelogContent.cs
Comment (Feodor0090) [reply osu.Game/Overlays/Changelog/ChangelogContent.cs]: I also had a question "why popover container isn't global". I added game-wide one inside `TooltipContainer`. Nothing seems to have broken from this.

Comment (peppy) [reply osu.Game/Overlays/Changelog/ChangelogContent.cs]: Bit hesitant to have that change bundled with this PR. At very least we'll want to go through all other usages of popovers as a follow-up and remove any local implementations that are no longer necessary (ie. anything that isn't in a scroll container and relying on the popover being localised to better track scroll).

**Comment by peppy** [comment]
So the immediate problem with moving the popover container to such a high level is it has the wrong colouring...

**Comment by peppy** [comment]
I've moved it to `OnlineOverlay`, I think this is okay for now.

**Review by peppy — APPROVED** [review]

**Pull request merged**
